### DISCUSSION TITLE
SamplingParameters doesn't have request information

### DISCUSSION THREAD

**Discussion by arabzade** [open-telemetry/opentelemetry-dotnet#4307] [Q&A] [answered]
Hi, Our team uses open telemetry for instrumentation of metrics and traces. Currently, we are populating all the traces (incoming/outgoing) requests but we would only populate traces that are interested and our decision would be based on the request url or request headers. For implementing that, we created a Custom sampler to filter out the traces that we are not interested in ShouldSample method, however sampler only gives us SamplingParameters which contains ParentContext, TraceID, Kind etc. but not request which I want to make decision from. Can you please help me figure out how can I get access to request information from SamplingParameters in ShouldSample method?

**Comment by utpilla** [comment, marked as answer]
The parameters required for `ShouldSample` are defined in the specification: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampler.

Is this an ASP .NET Core application? In that case, you could use the `Filter` option of AspNetCoreInstrumentation: https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.AspNetCore#filter

**Reply by cijothomas** [reply]
Sampler cannot be used to sample out based on url. unless the url is passed in the startActivity() call. (which is very unlikely in case of asp.net core case).

The above suggestion of Filter is your best bet, but it'll not be nice as the Sampler option.

**Reply by arabzade** [reply]
It's an ASP .NET Core application and I tried Filter option of AspNetCoreInstrumentation but still Filter option will be called after ShouldSample where sampling decision has already been made. This is how my TraceProvider look like:

tracerProvider = Sdk.CreateTracerProviderBuilder()
                    .AddSource("test")
                    .AddAspNetCoreInstrumentation((options) =>
                    {
                            options.Filter = (httpContext) =>
                            {
                               // retrieve required attributes  and add them to Activity to be used in our custom sampler for sampling decision
                                SetScenarioDimensions(httpContext.Request, out string scenarioId, out bool isDependency, out bool? dependencyExpectedSuccess);
                                Activity.Current.SetTag("ScenarioID", scenarioId);
                                return true;
                        };
                        options.EnrichWithHttpRequest = (activity, request) =>
                        {
                                SetScenarioDimensions(request, out string scenarioId, out bool isDependency, out bool? dependencyExpectedSuccess);
                                activity.SetTag("ScenarioID", scenarioId);
                        };
                    })
                    .AddHttpClientInstrumentation()
                    .AddGenevaTraceExporter(options =>
                    {
                              options.PrepopulatedFields = new Dictionary<string, object>
                              {
                                  ["cloud.role"] = this.cloudRole,
                                  ["cloud.roleInstance"] = this.cloudRoleInstance
                              };
                              if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                              {
                                  options.ConnectionString = "EtwSession=OpenTelemetry";
                              }
                              else
                              {
                                  options.ConnectionString = "Endpoint=unix:/var/run/mdsd/default_fluent.socket";
                              }
                    })
                .SetSampler(scenarioHealthSampler)
                .Build();

**Reply by utpilla** [reply]
> It's an ASP .NET Core application and I tried Filter option of AspNetCoreInstrumentation but still Filter option will be called after ShouldSample where sampling decision has already been made.

Yes, Filter option would be called after the sampling decision has been made but as Cijo already mentioned, this option is your best best.

The `Filter` Func that you have provided always returns true so every request would be processed. You want to return false for requests that you want to filter out. You could get hold of the url by using the `HttpContext` provided to you as the parameter to the Func.

**Reply by arabzade** [reply]
I am not trying to filter the requests but I am trying to filter only traces that are being sampled based on the requests. For example, if I see request http://test.com I want to drop the traces for that specific request and don't sample them. Is it possible to change the sampling decision in Filter Func based on the requests?

**Reply by utpilla** [reply]
Returning `false` for a given request in the `Filter` option would drop the trace for that particular request from being sent to any processors/exporters further down the chain.

**Reply by arabzade** [reply]
Thanks it worked!
